target_chains/starknet/contracts/src/reader.cairo

```diff
@@ -0,0 +1,333 @@
+use core::option::OptionTrait;
+use core::array::ArrayTrait;
+use core::keccak::cairo_keccak;
+use core::integer::u128_byte_reverse;
+use core::fmt::{Debug, Formatter};
+
+pub const EOF: felt252 = 'unexpected end of input';
+pub const UNEXPECTED_OVERFLOW: felt252 = 'unexpected overflow';
+
+/// A byte array with storage format similar to `core::ByteArray`, but
+/// suitable for reading data from it.
+#[derive(Drop, Clone, Serde)]
+pub struct ByteArray {
```

Comment:
AFAIK, the only way to read the data out of `core::ByteArray` is with `.at()`, only 1 byte at a time, so it would be inefficient. I created this utility so that we can slice data into larger chunks.

Reply:
Not so important but my comment was whether it is possible to implement the functionality that you've implemented here on core::ByteArray directly.